src/main/java/org/broadinstitute/hellbender/utils/genotyper/MoleculeLikelihoods.java

```diff
@@ -0,0 +1,1229 @@
+package org.broadinstitute.hellbender.utils.genotyper;
+
+import com.google.common.annotations.VisibleForTesting;
+import htsjdk.samtools.util.Locatable;
+import htsjdk.variant.variantcontext.Allele;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.objects.Object2IntArrayMap;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+import org.apache.commons.collections.ListUtils;
+import org.apache.commons.math3.stat.descriptive.rank.Median;
+import org.broadinstitute.hellbender.utils.IndexRange;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.pileup.PileupElement;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Read-likelihoods container implementation based on integer indexed arrays.
+ *
+ * @param <A> the type of the allele the likelihood makes reference to.
+ *
+ * Note: this class uses FastUtil collections for speed.
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class MoleculeLikelihoods<EVIDENCE extends Locatable, A extends Allele> implements SampleList, AlleleList<A> {
```

Comment:
I'm not loving this name, but TemplateLikelihoods or InsertLikelihoods seems weird too.  @yfarjoun any ideas?

Reply:
Ideas:
- `FragmentLikelihoods`
- `LinkedReadsLikelihoods'
- `PairedReadsLikelihoods` (though this might be more generic then paired reads)

Reply:
I want to be general enough to some day also deal with PCR duplicate sets in consensus reads.  So I'm trying to avoid assuming the evidence is just a pair of reads.  That said, I don't like my name, either.

Reply:
I suppose `LinkedReadsLikelihoods` does the job.

Reply:
I would simply drop the name of the evidence type from the class name... since it now supports any kind of evidence perhaps it should be a neutral name like "LikelihoodsTable" or "LikelihoodsMatrix"... there is likelihoods-matrix class already perhaps this one should be MultiSampleLikelihoodMatrix then. My point is that if the evidence type is a type-parameter the class name must not contain any the evidence class name even if that is the usual one.

Reply:
done